assets/js/googlesitekit/widgets/util/get-widget-layout.js

```diff
@@ -196,7 +202,18 @@ export function getWidgetLayout( widgets, widgetStates ) {
 			// If the column count without the overflowing widget is exactly 9, expand
 			// the widths of these widgets slightly to fill the entire 12 columns.
 			if ( counter === 9 ) {
-				[ classNames, columnWidths, counter ] = resizeClasses( classNames, columnWidths, counter );
+				// First zip the rows and columns
+				const rowColMap = getRowColumnMap( rowIndexes, columnWidths );
+				// Extract the columns on the current row to be resized (we
+				// don't want to include the overflowing widget or previously
+				// calculated rows)
+				let resizeRow = rowColMap[ rowIndex ];
+				// Resize them
+				[ resizeRow, counter ] = resizeColumns( resizeRow, counter );
+				// Insert the resized row back in
+				rowColMap[ rowIndex ] = resizeRow;
+				// Update the columnWidths with the resized row
+				columnWidths = Object.values( rowColMap ).flat();
```

Comment:
Why was the logic here altered and why a new function `getRowColumnMap` introduced? This should work as before, so why not just get rid of `classNames` and leave the other logic unaltered?

Reply:
My source of truth while working on this was the snapshots, the rationale being that if they all passed then I had a reasonable hope that the refactor didn't cause a regression. So I only deviated from the IB / ACs when the snapshot tests failed.

In this case, the resize operation seemed to malfunction and some snapshot tests started failing. For example this test:
```
WidgetAreaRenderer should resize widgets in a row that spans 9 columns to fill the full 12 columns 
(3, 12, 3-6, 6 -> 3, 12, 4-8, 6)
```
The resize operation calculated the 9-column row correctly (`3-6 -> 4-8`), but then continued into the previous rows. So here it adjusted the `3` on the first row to be a `4`, meaning the final output was `4, 12, 4-8, 6`.

I found that if I only passed the offending row, it was resized correctly and the tests passed.

I think it's to do with this change, because we lose the `includes` from the original, which might have been why we didn't see this issue with the previous code:
```js
-	if ( singleWidgetClassNames.includes( 'mdc-layout-grid__cell--span-3-desktop' ) ) {
-		singleWidgetClassNames[ singleWidgetClassNames.indexOf( 'mdc-layout-grid__cell--span-3-desktop' ) ] = 'mdc-layout-grid__cell--span-4-desktop';
+	if ( columnWidths[ i ] === 3 ) {
```

That's my best guess though - I'll be honest, I found this quite a tricky ticket! 😂
I'm very happy to remove this if the regression is not an issue. Or maybe I've misunderstood something along the way here.

Reply:
@johnPhillips I found the problem here. It was indeed broken before, but since previously we weren't using the column widths (nor really testing them) it never surfaced.

* The problem is https://github.com/google/site-kit-wp/pull/2927/files#diff-1accd10359d13467fb6f9186730cc4ed6a8b0aac9139453bcd5cb3a6f165e807R187
* In that line we add the column width for the current widget to the list of column widths before we have even fitted it into the current row. This is different from before where the class names were only added to the actual area at the very end (see https://github.com/google/site-kit-wp/pull/2927/files#diff-1accd10359d13467fb6f9186730cc4ed6a8b0aac9139453bcd5cb3a6f165e807L214, there's also a comment on why that was done).
* The problem is basically that due to adding the current column width to the list of widths too early, one more entry is in there than it should. Before this PR, this simply meant the column widths were out of sync with the class names passed to `resizeClasses`. Now that we actually rely on the column widths though, this surfaces as a problem, breaking the expected behavior.
* When reverting to the simpler previous code where it was only calling `resizeClasses` here and adding a `console.log` to the beginning of `resizeClasses`, you will see that it's always one too many numbers that is passed to the function. That becomes more apparent when you add the same `console.log` to the current `develop` codebase: In there, you'll be able to see that it's always one more entry in `columnWidths` than in `classNames`, and the last entry is the one that shouldn't be there.

In other words, TLDR:
* Please remove the new extra logic and function you've added here and rely on calling `resizeClasses` as before.
* Instead, to fix the problem remove the early addition to the `columnWidths` array in https://github.com/google/site-kit-wp/pull/2927/files#diff-1accd10359d13467fb6f9186730cc4ed6a8b0aac9139453bcd5cb3a6f165e807R187 and instead do that in https://github.com/google/site-kit-wp/pull/2927/files#diff-1accd10359d13467fb6f9186730cc4ed6a8b0aac9139453bcd5cb3a6f165e807R231.

Reply:
Ah, well spotted! I think that bug required an extra brain in the end 😂

I can't seem to get the link in your comment to work, but I think you meant [this line](https://github.com/google/site-kit-wp/blob/develop/assets/js/googlesitekit/widgets/util/get-widget-layout.js#L181)?
```js
// Populate initial column width and row index for the widget.
// These may be corrected below in certain situations.
columnWidths.push( WIDTH_GRID_COUNTER_MAP[ width ] ); //  ← 💥
rowIndexes.push( rowIndex );
```

When I remove that, the tests pass, so hopefully that's it. Let me know if I misunderstood.

I've removed the extra logic too 👍